rfd/0098-registered-openssh-nodes.md

```diff
@@ -0,0 +1,106 @@
+
+## Details
+
+### Registering nodes
+
+A new sub-kind to the `node` resource will be added for registered OpenSSH nodes: `openssh`. The absence of a `node` resource sub-kind (or presence of a `teleport` sub-kind) will imply that a node is a Teleport agent node, making this change backwards compatible.
+
+#### Registering with `tctl`
+
+`tctl` should not require as many fields to be set when creating nodes. This is an example `node` resource that will work with `tctl create --force` today:
+
+```yaml
+kind: node
+metadata:
+  name: 5da56852-2adb-4540-a37c-80790203f6a9
+spec:
+  addr: 1.2.3.4:22
+  hostname: agentless-node
+version: v2
+```
+
+`tctl` will auto-generate `metadata.name` if it is not already set so users don't have to generate GUIDs themselves if `sub_kind` is `openssh`. Also, if `sub_kind` is set to `openssh`, `spec.public_addr` will not be allowed for registered OpenSSH nodes as it is not needed. An example of a registered OpenSSH node resource:
+
+```yaml
+kind: node
+sub_kind: openssh
+metadata:
+  name: 5da56852-2adb-4540-a37c-80790203f6a9
+spec:
+  addr: 1.2.3.4:22
+  hostname: agentless-node
+version: v2
+```
+
+When manually registering an OpenSSH node, create a registered OpenSSH node resource file and create the resource in the cluster with `tctl create --force <node.yml>`. Then follow the steps outlined below in `Manual certificate rotation` to configure the registered OpenSSH node.
+
+### OpenSSH CA
+
+For security related reasons that will be discussed below, a new CA will be added called OpenSSH CA. The OpenSSH CA will be responsible for generating and signing user certificates that will be used to authenticate with registered OpenSSH nodes. The public key of the OpenSSH CA will be copied to all registered OpenSSH nodes and configured as `TrustedUserCAKeys` in `sshd_config`.
+
+#### Manual certificate rotation
+
+Users not using EC2 discovery mode and hosting registered OpenSSH nodes themselves will have to rotate the OpenSSH CA manually and generate new SSH host certificates. They can do so by:
+
+1. Exporting the OpenSSH CA: `tctl auth export --type=openssh | sed "s/cert-authority\ //"`
+2. Write the exported CA to the configuration folder of the node's SSH daemon, ex. `/etc/ssh/` for OpenSSH on Linux
+3. Replace the `TrustedUserCAKeys` setting in `sshd_config` to point to the newly created CA
+4. Generate a new host certificate: `tctl auth sign --host=<addr,hostname> --format=openssh --out=host-cert`
+5. Replace the existing host certificate with the new one on the node and set `HostKey` and `HostCertificate` in `sshd_config`
+
+### RBAC
+
+When a user sends a request to a Proxy to connect to a node, the Proxy will attempt to find a node resource by either its hostname or IP, whichever the user specified. If the resource exists and has the `openssh` `sub_kind`, an RBAC check will be performed. If the resource does not exist or isn't an `openssh` node, the connection flow will continue as normal.
```

Comment:
Isn't this concern alleviated by using a different CA for OpenSSH vs Teleport nodes? Off the top of my head I'm not sure what extra string check buys us tbh.

Reply:
@jakule are you saying the server string check will allow us to report a more useful error to the client?

Reply:
I'm more concerned about registering OpenSSH node as Teleport and skipping the RBAC check, but even then, the user will get an authentication error, so from a security perspective, we're "fine". From a usability perspective, we could ping each server and check the client string, but this would make the registration slow and would require all nodes to be available. I think that the current compromise is fine.